File: src/lib/core/llm/anthropic.ts
import { fetchWithRetry } from "../http";
import type { LLMCompleteOptions, LLMProvider } from "./types";

const URL = "https://api.anthropic.com/v1/messages";

interface AnthropicTextBlock {
  type?: string;
  text?: string;
}

/** Anthropic (Claude) Messages API adapter. Uses raw fetch — no SDK dependency. */
export class AnthropicLLM implements LLMProvider {
  readonly name = "anthropic";
  readonly model: string;
  private readonly apiKey: string;

  constructor(apiKey: string, model?: string) {
    this.apiKey = apiKey;
    this.model = model || process.env.ANTHROPIC_MODEL || "claude-haiku-4-5-20251001";
  }

  async complete(prompt: string, opts: LLMCompleteOptions = {}): Promise<string> {
    const system =
      (opts.system ?? "") +
      (opts.json ? "\nRespond with ONLY a single valid JSON object — no prose, no markdown fences." : "");

    const res = await fetchWithRetry(
      URL,
      {
        method: "POST",
        headers: {
          "x-api-key": this.apiKey,
          "anthropic-version": "2023-06-01",
          "content-type": "application/json",
        },
        body: JSON.stringify({
          model: opts.model || this.model,
          max_tokens: opts.maxTokens ?? 1024,
          temperature: opts.temperature ?? 0.3,
          ...(system.trim() ? { system: system.trim() } : {}),
          messages: [{ role: "user", content: prompt }],
        }),
      },
      { timeoutMs: opts.timeoutMs ?? 30000 }
    );

    const json = await res.json();
    const blocks: AnthropicTextBlock[] = Array.isArray(json?.content) ? json.content : [];
    return blocks
      .filter((b) => b?.type === "text")
      .map((b) => b.text || "")
      .join("")
      .trim();
  }
}

File: worker_py/llm/anthropic.py
"""Anthropic (Claude) Messages API adapter. Port of ``llm/anthropic.ts``."""

import os
from typing import Optional

from ..http import request_with_retry
from .base import LLMProvider

URL = "https://api.anthropic.com/v1/messages"


class AnthropicLLM(LLMProvider):
    name = "anthropic"

    def __init__(self, api_key: str, model: Optional[str] = None) -> None:
        self.api_key = api_key
        self.model = model or os.environ.get("ANTHROPIC_MODEL") or "claude-haiku-4-5-20251001"

    async def complete(
        self,
        prompt,
        *,
        system=None,
        max_tokens=1024,
        temperature=0.3,
        json=False,
        model=None,
        timeout_ms=30000,
    ) -> str:
        sys = (system or "") + (
            "\nRespond with ONLY a single valid JSON object — no prose, no markdown fences."
            if json
            else ""
        )
        body = {
            "model": model or self.model,
            "max_tokens": max_tokens,
            "temperature": temperature,
            "messages": [{"role": "user", "content": prompt}],
        }
        if sys.strip():
            body["system"] = sys.strip()

        res = await request_with_retry(
            "POST",
            URL,
            headers={
                "x-api-key": self.api_key,
                "anthropic-version": "2023-06-01",
                "content-type": "application/json",
            },
            json=body,
            timeout_ms=timeout_ms,
        )
        data = res.json()
        blocks = data.get("content") if isinstance(data, dict) else None
        if not isinstance(blocks, list):
            return ""
        return "".join(b.get("text", "") for b in blocks if b.get("type") == "text").strip()
